csrc/multidevice/utils.cpp

```diff
@@ -240,17 +267,32 @@ void insertReshardings(Fusion* fusion) {
         "multi-output expressions are not supported");
     auto output = expr->outputs().at(0)->as<TensorView>();
     std::vector<TensorView*> new_inputs;
-    for (auto input : getTvsWithDifferentSharding(
-             output, ir_utils::filterByType<TensorView>(expr->inputs()))) {
-      // TODO: reuse cacheAfter?
-      // TODO: here we should add a mechanism to potentially reuse the inserted
-      // resharding accross all the consumer of the resharded tensor. This way
-      // we could avoid wasteful resharding set insertion.
-      TensorView* new_input = set(input);
-      new_inputs.push_back(new_input);
-      expr = ir_utils::replaceValInExprInputs(expr, input, new_input);
+    auto inputs = getTvsWithDifferentSharding(
+        output, ir_utils::filterByType<TensorView>(expr->inputs()));
+    // Insert resharding expression after the expr when there is only one input.
+    // input [expr] output [set] new_output
+    if (!inputs.empty() && expr->inputs().size() == 1) {
```

Comment:
The motivation was just simplicity for this PR. For an expression with more than input, if we add the resharding after the expr, we will have to decide which input we want to shardLike and potentially reshard some inputs, because the inputs might not have consistent sharding.